test/intrusive_list_properties.cpp

```diff
@@ -0,0 +1,143 @@
+#include <stdx/intrusive_list.hpp>
+
+#include <catch2/catch_test_macros.hpp>
+#include <rapidcheck/catch.h>
+#include <rapidcheck/state.h>
+
+#include <list>
+
+namespace {
+struct int_node {
+    int value{};
+    int_node *prev{};
+    int_node *next{};
+};
+} // namespace
+
+using ListModel = std::list<int>;
+
+template <typename ListSut> struct IntrusiveListCommands {
+    static void rc_assert_equal(ListModel const &m, ListSut const &sut) {
+        RC_ASSERT(m.empty() == sut.empty());
+
+        auto mi = m.cbegin();
+        auto si = sut.cbegin();
+
+        while (mi != m.cend() && si != sut.cend()) {
+            RC_ASSERT(*mi == si->value);
+            mi++;
+            si++;
+        }
+
+        RC_ASSERT(mi == m.cend());
+        RC_ASSERT(si == sut.cend());
+    }
+
+    struct PushBack : rc::state::Command<ListModel, ListSut> {
+        int value = *rc::gen::inRange(0, 100);
+
+        void checkPreconditions(ListModel const &) const override {}
+
+        void apply(ListModel &m) const override { m.push_back(value); }
+
+        void run(ListModel const &m, ListSut &sut) const override {
+            rc_assert_equal(m, sut);
+            sut.push_back(new int_node{value});
+        }
+
+        void show(std::ostream &os) const override {
+            os << "push_back(" << value << ")";
+        }
+    };
+
+    struct PopBack : rc::state::Command<ListModel, ListSut> {
+        void checkPreconditions(ListModel const &m) const override {
+            RC_PRE(!m.empty());
+        }
+
+        void apply(ListModel &m) const override { m.pop_back(); }
+
+        void run(ListModel const &m, ListSut &sut) const override {
+            rc_assert_equal(m, sut);
+            delete sut.pop_back();
+        }
+
+        void show(std::ostream &os) const override { os << "pop_back()"; }
+    };
+
+    struct PushFront : rc::state::Command<ListModel, ListSut> {
+        int value = *rc::gen::inRange(0, 100);
+
+        void checkPreconditions(ListModel const &) const override {}
+
+        void apply(ListModel &m) const override { m.push_front(value); }
+
+        void run(ListModel const &m, ListSut &sut) const override {
+            rc_assert_equal(m, sut);
+            sut.push_front(new int_node{value});
+        }
+
+        void show(std::ostream &os) const override {
+            os << "push_front(" << value << ")";
+        }
+    };
+
+    struct PopFront : rc::state::Command<ListModel, ListSut> {
+        void checkPreconditions(ListModel const &m) const override {
+            RC_PRE(!m.empty());
+        }
+
+        void apply(ListModel &m) const override { m.pop_front(); }
+
+        void run(ListModel const &m, ListSut &sut) const override {
+            rc_assert_equal(m, sut);
+            delete sut.pop_front();
+        }
+
+        void show(std::ostream &os) const override { os << "pop_front()"; }
+    };
+};
+
+TEST_CASE("push_back, pop_front", "[intrusive_list_properties]") {
```

Comment:
**question**: Where are the intrusive list nodes kept? Because the list doesn't own them...

Reply:
they are being dynamically allocated and freed...this is related to why the address sanitizer is failing. 😭 i will need a simple scheme to free all the nodes when a list goes out of scope.